tasks/eva_2124/correct_contig_for_study.py

```diff
@@ -0,0 +1,171 @@
+
+    if contig in by_genbank:
+        return contig, True
+
+    raise Exception('could not find synonym for contig {}'.format(contig))
+
+
+def get_mongo_connection_handle_url(host, port=27017, username=None, password=None, authentication_database="admin") -> pymongo.MongoClient:
+    mongo_connection_uri = "mongodb://"
+    if username and password:
+        mongo_connection_uri += '%s:%s@' % (username, quote_plus(password))
+    mongo_connection_uri += '%s:%s/%s' % (host, port, authentication_database)
+    return pymongo.MongoClient(mongo_connection_uri)
+
+
+def correct(mongo_user, mongo_password, mongo_host, studies, assembly_accession, mongo_database, assembly_report=None,
+            chunk_size=1000):
+    """
+    Connect to mongodb and retrieve all variants the should be updated, check their key and update them in bulk.
+    """
+    with get_mongo_connection_handle_url(
+            username=mongo_user,
+            password=mongo_password,
+            host=mongo_host
+    ) as accessioning_mongo_handle:
+        sve_collection = accessioning_mongo_handle[mongo_database]["submittedVariantEntity"]
+        logging.info("Loading synonyms...")
+        synonym_dictionaries = load_synonyms_for_assembly(assembly_accession, assembly_report)
+        assert_all_contigs_can_be_replaced(sve_collection, synonym_dictionaries, studies, assembly_accession)
+        return do_updates(sve_collection, synonym_dictionaries, studies, assembly_accession, chunk_size)
+
+
+def assert_all_contigs_can_be_replaced(sve_collection, synonym_dictionaries, studies, assembly_accession):
+    logging.info("Checking that all contigs are replaceable...")
+    cursor = sve_collection.aggregate([{'$match': {'seq': assembly_accession, 'study': {'$in': studies}}},
+                                       {'$group': {'_id': '$contig', 'count': {'$sum': 1}}}])
+    already_genbank_contigs = 0
+    already_genbank_variants = 0
+    replaceable_contigs = 0
+    replaceable_variants = 0
+    unreplaceable_variants = 0
+    unreplaceable_contigs = set()
+    for contig in cursor:
+        try:
+            genbank, was_already_genbank = get_genbank(synonym_dictionaries, contig['_id'])
+            if was_already_genbank:
+                already_genbank_contigs += 1
+                already_genbank_variants += contig['count']
+            else:
+                replaceable_contigs += 1
+                replaceable_variants += contig['count']
+        except Exception:
+            unreplaceable_variants += contig['count']
+            unreplaceable_contigs.add(contig['_id'])
+
+    if len(unreplaceable_contigs) > 0:
+        raise Exception(
+            'Aborting replacement (no changes were done).'
+            ' With the provided assembly_report, the next {} contigs (present in {} variants) can not be replaced: {}'
+            .format(len(unreplaceable_contigs), unreplaceable_variants, unreplaceable_contigs))
+    elif replaceable_contigs == 0:
```

Comment:
Shouldn't this be replaceable_contigs == 0 && unreplaceable_contigs == 0 && len(already_genbank_contigs) > 0 ?

Reply:
the current condition is the same as your suggestion but also including the case of `replaceable_contigs == 0 && already_genbank_contigs == 0`, which means that there are no variants in the DB (because we have also checked that `len(unreplaceable_contigs) == 0`), so arguably we should also do the exception, abort and message questioning the parameters.